KEYCLOAK-11908 Support for conditional creating indices based on number

y-tabata opened the pull request:
JIRA ticket: https://issues.redhat.com/browse/KEYCLOAK-11908

# How to use

- Change indexCreationThreshold property of connectionsLiquibase spi in the configuration file (like standalone.xml).

  ```xml
              <spi name="connectionsLiquibase">
                  <provider name="default" enabled="true">
                      <properties>
                          <property name="indexCreationThreshold" value="10000"/>
                      </properties>
                  </provider>
              </spi>
  ```

- Define createIndex change to the change log file (jpa-changelog-x.x.x.xml).

  ```xml
      <changeSet author="keycloak" id="add-index">
          <createIndex indexName="xxx" tableName="yyy">
              <column name="zzz" type="VARCHAR(255)"/>
          </createIndex>
      </changeSet>
  ```

- When migration, if the number of records of the specified table is larger than indexCreationThreshold, createIndex change does not run, otherwise createIndex change run.



# Spec

- The indexCreationThreshold value is common in all tables.
- The default indexCreationThreshold value is 10000.
- If the indexCreationThreshold value is -1, all createIndex change run.



# Code

- `DefaultLiquibaseConnectionProvider`
    - new field `indexCreationThreshold`
        - In `init` method, set this value from the indexCreationThreshold property's value in the configuration file.
        - In `getLiquibase` method, pass `indexCreationThreshold` to `LiquibaseJpaUpdaterProvider` and `QuarkusJpaUpdaterProvider` via Liquibase's changeLogParameter.
-  `LiquibaseJpaUpdaterProvider` and `QuarkusJpaUpdaterProvider`
    - new `validateIndexThreshold` method
        - If `indexCreationThreshold`  is not defined, or `indexCreationThreshold` equals -1, skip verification.
        - If there is a change instance of `CreateIndexChange` in the changeSet of unrunChangeSets,
            - if the target table exists and the number of records of the target table is larger than `indexCreationThreshold`, mark the changeSet as `validationFailed`.
    - In `updateChangeSet` method, call `validateIndexThreshold` method when there are unrunChangeSets.

y-tabata committed:
KEYCLOAK-11908 Support for conditional creating indices based on number
of records

y-tabata commented:
@hmlnarik Could you check this?

vramik reviewed (changes requested):
@y-tabata thank you very much for your contribution. 

If I looked correctly, when the `validationFailed` switch of the `ChangeSet` is set to `true` it'd mean whole changeset won't be executed. There could be more than one change in a changeset, not just `createIndex`, so we cannot use this approach. 

We'd suggest to create `CustomCreateIndexChange` which will extend `liquibase.change.core.CreateIndexChange` and override `generateStatements(Database)` method where you could decide (based on `DefaultLiquibaseConnectionProvider.indexCreationThreshold`) if you will print the sql in the logs or use `super.generateStatements(database)`. The `CustomCreateIndexChange` could be then registered in `DefaultLiquibaseConnectionProvider.baseLiquibaseInitialization()` using `ChangeFactory.getInstance().register(new CustomCreateIndexChange());`.

Please note there has to be higher priority set for the `CustomCreateIndexChange`, please see `liquibase.change.DatabaseChange` annotation. 

There is also manual mode [1] for migration, where keycloak creates sql script for you instead of performing migration directly to the database. The PR must ensure that all indexes are created regardless of the `indexCreationThreshold` in case of manual migration.

I'd suggest not to add the cli scripts because you'd have to update default configuration of xml's in wildfly as well (not sure about Quarkus). I think we can safely count on default value you've introduced: `indexCreationThreshold = config.getInt("indexCreationThreshold", 10000);`. The admin will still have a possibility to add the `<spi name="connectionsLiquibase">` with his preffered value. The documentation PR will be needed to describe this parameter.

[1] https://www.keycloak.org/docs/latest/upgrading/index.html#manual-relational-database-migration

y-tabata commented:
@vramik thank you for reviewing.
> We'd suggest to create CustomCreateIndexChange which will extend liquibase.change.core.CreateIndexChange and override generateStatements(Database) method where you could decide (based on DefaultLiquibaseConnectionProvider.indexCreationThreshold) if you will print the sql in the logs or use super.generateStatements(database).

Done.

> The CustomCreateIndexChange could be then registered in DefaultLiquibaseConnectionProvider.baseLiquibaseInitialization() using ChangeFactory.getInstance().register(new CustomCreateIndexChange());.

Done but registered `CustomCreateIndexChange.class`.

> Please note there has to be higher priority set for the CustomCreateIndexChange, please see liquibase.change.DatabaseChange annotation.

Set `changeMetaData.getPriority() + 1` to the priority in `createChangeMetaData()`.

> There is also manual mode [1] for migration, where keycloak creates sql script for you instead of performing migration directly to the database. The PR must ensure that all indexes are created regardless of the indexCreationThreshold in case of manual migration.

Added `LoggingExecutor` check in `generateStatements(Database database)`.

> I'd suggest not to add the cli scripts because you'd have to update default configuration of xml's in wildfly as well (not sure about Quarkus). I think we can safely count on default value you've introduced: indexCreationThreshold = config.getInt("indexCreationThreshold", 10000);.

Deleted the cli scripts.

> The admin will still have a possibility to add the <spi name="connectionsLiquibase"> with his preffered value. The documentation PR will be needed to describe this parameter.

I will prepare it.

y-tabata commented:
Docs PR: https://github.com/keycloak/keycloak-documentation/pull/1072

vramik reviewed (changes requested):
Thank you @y-tabata for quick update. I have several comments, please see inline. 

Please ensure that the functionality from this PR is tested, ideally with Postgres and a different schema, before submitting a new version.

y-tabata commented:
@vramik 
> If I looked correctly, when the validationFailed switch of the ChangeSet is set to true it'd mean whole changeset won't be executed. There could be more than one change in a changeset, not just createIndex, so we cannot use this approach.

Liquibase manages whether migration executed or not per ChangeSet with "databasechangelog" table. And to change from per the ChangeSet to per Change needs to change Liquibase, I think.

To achieve "Should the index be created by administrator, the message needs to be logged every time Keycloak starts.", we need to set the unrun status or delete the run status for the ChangeSet every time Keycloak starts. To set the unrun status (FAILED or SKIPPED) is not supported, so we need to delete the run status.

When a CreateIndexChange skipped and the ChangeSet marked as "Run", the CreateIndexChange will be never executed after that.
When a CreateIndexChange skipped and the ChangeSet marked as "UnRun", the changes except for the CreateIndexChange will be failed in the next Keycloak startup and Keycloak startup also fails.

I added a check that if the ChangeSet includes only one Change or not.
From now on, if we add the index which seems to affect Keycloak's performance, we need to define only one Change in one ChangeSet, is it acceptable?

y-tabata commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/CustomCreateIndexChange.java:
```diff
@@ -0,0 +1,80 @@
+package org.keycloak.connections.jpa.updater.liquibase.custom;
```
Deleted CustomCreateIndexChange.

y-tabata commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/CustomCreateIndexChange.java:
```diff
@@ -0,0 +1,80 @@
+
+        try {
+            if (!SnapshotGeneratorFactory.getInstance()
+                .has(new Table().setName(getTableName()).setSchema(new Schema(getCatalogName(), getSchemaName())), database))
```
If this check not exists, the initial startup fails.

y-tabata replied:
Added a comment.

y-tabata commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/CustomCreateIndexChange.java:
```diff
@@ -0,0 +1,80 @@
+                return super.generateStatements(database);
+
+            int result = ExecutorService.getInstance().getExecutor(database)
+                .queryForInt(new RawSqlStatement("select count(*) from" + getTableName()));
```
Modified. And I tested on PostgreSQL.

y-tabata commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/CustomCreateIndexChange.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @Override
+    public ChangeMetaData createChangeMetaData() {
```
Deleted CustomCreateIndexChange.

y-tabata commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/conn/DefaultLiquibaseConnectionProvider.java:
```diff
@@ -109,12 +116,14 @@ protected void baseLiquibaseInitialization() {
         SqlGeneratorFactory.getInstance().register(new CustomLockDatabaseChangeLogGenerator());
-    }
 
+        ChangeFactory.getInstance().register(CustomCreateIndexChange.class);
```
Deleted CustomCreateIndexChange.

y-tabata commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/CustomCreateIndexChange.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public SqlStatement[] generateStatements(Database database) {
+        if (ExecutorService.getInstance().getExecutor(database) instanceof LoggingExecutor)
```
Deleted CustomCreateIndexChange.

y-tabata commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/conn/DefaultLiquibaseConnectionProvider.java:
```diff
@@ -109,12 +116,14 @@ protected void baseLiquibaseInitialization() {
     @Override
     public void init(Config.Scope config) {
-
+        indexCreationThreshold = config.getInt("indexCreationThreshold", 10000);
```
Changed.

y-tabata commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/CustomCreateIndexChange.java:
```diff
@@ -0,0 +1,80 @@
+        Object indexCreationThreshold = ((AbstractJdbcDatabase) database)
+            .get(DefaultLiquibaseConnectionProvider.INDEX_CREATION_THRESHOLD_PARAM);
+
+        if (indexCreationThreshold == null || (Integer) indexCreationThreshold == -1)
```
Changed.

hmlnarik commented:
Thank you @y-tabata.

> To achieve "Should the index be created by administrator, the message needs to be logged every time Keycloak starts.", we need to set the unrun status or delete the run status for the ChangeSet every time Keycloak starts. To set the unrun status (FAILED or SKIPPED) is not supported, so we need to delete the run status.

This is correct, and we should not change the statuses. 

The intention here is to make sure that the admin is for sure notified that they are running on degraded-performance installation. If only a single warning is logged and then never again, it could be easy for the admin to forget to create the index manually.

I wonder whether there is a way to only inspect the changelog and check for createIndex changes? If not, then let's log a warning only once for now. There might be a way to record this failure somewhere (is it possible to do so directly in liquibase changelog?), and upon start validate that the recorded indices have been created already. That could be done as a follow-up task though.

> From now on, if we add the index which seems to affect Keycloak's performance, we need to define only one Change in one ChangeSet, is it acceptable?

@y-tabata I am afraid think this is not viable, the `createIndex` should be still possible to put into a changeset with several changes. However, if no statement is generated by the `CreateIndexChange` and instead a warning is logged when the table is too big, the functionality would be there.

y-tabata commented:
Thank you @hmlnarik.

> I wonder whether there is a way to only inspect the changelog and check for createIndex changes? If not, then let's log a warning only once for now. There might be a way to record this failure somewhere (is it possible to do so directly in liquibase changelog?), and upon start validate that the recorded indices have been created already. That could be done as a follow-up task though.

I changed to log a warning only once. And I changed that unrun Change is written in `comments` of the `databasechangelog` record.

vramik commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/CustomCreateIndexChange.java:
```diff
@@ -0,0 +1,188 @@
+        if (ExecutorService.getInstance().getExecutor(database) instanceof LoggingExecutor)
+            return super.generateStatements(database);
+
+        Object indexCreationThreshold = ((AbstractJdbcDatabase) database)
```
What'd you say about idea to create make `DefaultLiquibaseConnectionProvider.indexCreationThreshold` `static` and to introduce 
```
    public static int getIndexCreationThreshold() {
        return indexCreationThreshold;
    }
```
there? We'd then be able to use 
```
if (DefaultLiquibaseConnectionProvider.getIndexCreationThreshold() <= 0)
                return super.generateStatements(database);
``` 
here instead of parsing `Integer` from attributes.

hmlnarik replied:
I'd say if the `indexCreationThreshold` would need to be static, then the current approach is cleaner.

y-tabata replied:
@vramik @hmlnarik
I keep this implementation as-is.

vramik commented on model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/CustomCreateIndexChange.java:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    @Override
+    public boolean generateStatementsVolatile(Database database) {
```
Can we avoid overriding this and following two methods?

y-tabata replied:
Without these 3 override methods and without expressly calling the `generateStatements` method of the superclass, the `generateStatements` method of this class is called 14 times in one migration (so the warning outputs 14 times).
I also considered the way to add if which judges the caller method by using stack trace in the `generateStatements` method, but I think the current way is smarter.
If there is a better way, I'm happy if you tell me.

vramik reviewed (commented):
Thank you @y-tabata for the update. Great job. Please see my my comments inline.

vramik reviewed (dismissed):
@y-tabata I wanted to let you know I happy with the code as it is right now. I've tested it and works as expected. Please consider my comments above as suggestions or questions. I wanted to know your opinion on it.

vramik reviewed (approved):
Thank you @y-tabata for your work. Great job.

hmlnarik reviewed (approved):
@y-tabata Thank you for the contribution! I'm approving based on @vramik's approval. Also I'm not merging it yet, will merge once the docs PR would be reviewed.
